request-controller/src/test/java/org/wildfly/extension/requestcontroller/WFCORE4967_TestCase.java

```diff
@@ -0,0 +1,63 @@
+package org.wildfly.extension.requestcontroller;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.Test;
+
+public class WFCORE4967_TestCase {
+    private static final int TASKS_QTY = 10;
+    private static final int THREADS_QTY = 20;
+
+    @Test
+    public void noQueuedTasksLossWhenRunningRequestCompleteOnSuspendedRC() throws InterruptedException {
+        final AtomicInteger executedTaskCount = new AtomicInteger();
+
+        RequestController rc = suspendedRCWithQueuedTasks(TASKS_QTY, () -> {
+            executedTaskCount.incrementAndGet();
+        });
+
+        CountDownLatch latch = new CountDownLatch(THREADS_QTY);
+        createSynchronisedThreads(latch, () -> {
+            rc.requestComplete();
+        }).forEach(Thread::start);
+        latch.await();
+
+        rc.resume();

```

Comment:
So AIUI the basic problem is that one or more of the threads that called rc.requestComplete() when paused == true  is executing findForcedTask and has drained some tasks from taskQueue into the 'storage' list. And at the same time rc.resume is called and it therefore finds the queue empty and returns.  When the thread in findForcedTask puts its 'storage' tasks back in the queue there is nothing to execute them.

Is that a correct summary?

Reply:
Correct but not full in-depth. This tmp 'storage' list is problematic in multithreading context.
It causes problems before `rc.resume()` is called as well. 
We can get 
```
java.lang.NullPointerException
	at org.wildfly.extension.requestcontroller.RequestController$QueuedTask.access$1(RequestController.java:457)
	at org.wildfly.extension.requestcontroller.RequestController.findForcedTask(RequestController.java:404)
	at org.wildfly.extension.requestcontroller.RequestController.runQueuedTask(RequestController.java:386)
	at org.wildfly.extension.requestcontroller.RequestController.requestComplete(RequestController.java:238)
```
as an example

Reply:
@khroolick.  Yes, I saw the NPE potential and was hoping preventing that would be sufficient to fix the bug without adding the synchronized.  But your test quickly demonstrated that wasn't the case.

The synchronized makes me nervous because of possible interaction between the current synchronization on the RequestController object, which involve the server management layer manipulating the RC, and this new one which involves users of the RC queuing or completing tasks. For example the 'suspended' method holds the lock while the 'requestCountListener.done()' callback does its thing, and meanwhile threads trying to queue/complete tasks are blocking.

TBH that concern might be excessive.

I thought about instead synchronizing on the taskQueue where necessary but I don't know if that is better. With that you've got two separate locks being taken in some cases, which can have its own problems.  I didn't think hard about it though. WDYT?

Reply:
I had similar thinking, but I was not afraid of these operations as in our case we do not use them often. I didn't think too hard as well and just went with “simplest” fix I could apply.

In fact, if there is a fear of unnecessary blocking, we can improve the implementation of the `findForcedTask` method and avoid synchronization. It looks good after static analysis.

- If a thread has drained some tasks from taskQueue into the 'storage' list when `resume ()` has been called, these tasks will return to `taskQueue` and will be started under `requestComplete()` invocations. 
The only risk I see: Very unusual coincidence in which there will be too few calls requestComplete after server is resumed and timers schedule will be delayed until server will receive some traffic.

This is very very unlikely if not possible at all, as if we are already in this scenario of high concurrent access to `taskQueue` via` findForcedTask`.

Reply:
- I've extended unit test to simulate some additional requests after `rc.resume()` is called. 
- Tested above change with both 
 o [minimal example project](https://github.com/guestlinelabs/wildfly-timer-issue)
 o our complex EAR application

LGTM, please review this change again @bstansberry

Reply:
I'm unresolving this just so the discussion stays visible in the github ui as it may be useful in the future.